sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -673,4 +772,64 @@ private void removeOne(String filename) throws IOException {
       }
     }
   }
+
+  /**
+   * Implementations create instances of {@link WritableByteChannel} used by {@link FileBasedSink}
+   * and related classes to allow <em>decorating</em>, or otherwise transforming, the raw data that
+   * would normally be written directly to the {@link WritableByteChannel} passed into
+   * {@link WritableByteChannelFactory#create(WritableByteChannel)}.
+   */
+  public interface WritableByteChannelFactory extends Serializable {
+    /**
+     * @param channel the {@link WritableByteChannel} to wrap
+     * @return the {@link WritableByteChannel} to be used during output
+     * @throws IOException
+     */
+    public WritableByteChannel create(WritableByteChannel channel) throws IOException;
+
+    /**
+     * @return the MIME type that should be used for the files that will hold the output data
+     * @see MimeTypes
+     * @see <a href=
+     *      'http://www.iana.org/assignments/media-types/media-types.xhtml'>http://www.iana.org/assignments/media-types/media-types.xhtml</a>
+     */
+    public String getMimeType();
+
+    /**
+     * @return an optional filename suffix, eg, ".gz" is returned by {@link CompressionType#GZIP}
+     */
+    @Nullable
+    public String getFilenameSuffix();
+  }
+
+  /**
+   * Convenience implementation. Implementors of more complex custom
+   * {@link WritableByteChannelFactory}s can extend this class and just override, eg, the
+   * {@link #write(ByteBuffer)} method.
+   *
+   * <p>All methods of {@link WritableByteChannel} simply delegate to the given
+   * {@link WritableByteChannel}.
+   */
+  public static class DelegatingWritableByteChannel implements WritableByteChannel {
```

Comment:
Sure.  I left this there for easier subclassing of WritableByteChannel, but yeah, probably YAGNI.